src/battle_4.c

```diff
@@ -0,0 +1,18534 @@
+void sub_814A880(u8 a1, u8 a2);
+u8 sub_8015A98(u8 bank, u8 unusable_moves, u8 flags); //choose move limitations
+void sub_801529C(u8 bank);
+bool8 IsLinkDoubleBattle(void);
+void sub_8094B6C(u8 bank, u8 partyID, u8 r2);
+
+//extern BattleScripts
+extern u8 BattleScript_EndTurn[];
+extern u8 BattleScript_NoPPForMove[];
+extern u8 BattleScript_MagicCoatBounce[];
+extern u8 BattleScript_TookAttack[];
+extern u8 BattleScript_SnatchedMove[];
+extern u8 BattleScript_Pausex20[];
+extern u8 BattleScript_SubstituteFade[];
+extern u8 BattleScript_HangedOnMsg[];
+extern u8 BattleScript_OneHitKOMsg[];
+extern u8 BattleScript_EnduredMsg[];
+extern u8 BattleScript_PSNPrevention[];
+extern u8 BattleScript_BRNPrevention[];
+extern u8 BattleScript_PRLZPrevention[];
+extern u8 BattleScript_FlinchPrevention[];
+extern u8 BattleScript_StatUp[];
+extern u8 BattleScript_StatDown[];
+extern u8 BattleScript_NoItemSteal[];
+extern u8 BattleScript_ItemSteal[];
+extern u8 BattleScript_RapidSpinAway[];
+extern u8 BattleScript_TargetPRLZHeal[];
+extern u8 BattleScript_KnockedOff[];
+extern u8 BattleScript_LevelUp[];
+extern u8 BattleScript_WrapFree[];
+extern u8 BattleScript_LeechSeedFree[];
+extern u8 BattleScript_SpikesFree[];
+extern u8 BattleScript_ButItFailed[];
+extern u8 BattleScript_ObliviousPreventsAttraction[];
+extern u8 BattleScript_MistProtected[];
+extern u8 BattleScript_AbilityNoStatLoss[];
+extern u8 BattleScript_AbilityNoSpecificStatLoss[];
+extern u8 BattleScript_TrainerBallBlock[];
+extern u8 BattleScript_WallyBallThrow[];
+extern u8 BattleScript_SuccessBallThrow[];
+extern u8 BattleScript_ShakeBallThrow[];
+
+extern u8 gUnknown_081D919F[]; //spikes1
+extern u8 gUnknown_081D9171[]; //spikes2
+extern u8 gUnknown_081D91CD[]; //spikes3
+extern u8 BattleScript_1D6F44[]; //present dmg
+extern u8 BattleScript_1D83B5[]; //present full hp
+extern u8 BattleScript_1D839B[]; //present hp heal
+extern u8 BattleScript_1D6F74[];
+extern u8 gUnknown_081D977D[]; //castform change bs
+extern u8 gUnknown_081D9834[];
+extern u8 gUnknown_081D90FC[]; //bs random switchout
+extern u8 gUnknown_081D95DB[]; //bs payday money give
+
+//useful macros
+//read via orr
+#define BSScriptRead32(ptr) ((ptr)[0] | (ptr)[1] << 8 | (ptr)[2] << 16 | (ptr)[3] << 24)
+#define BSScriptRead16(ptr) ((ptr)[0] | (ptr)[1] << 8)
+#define BSScriptRead8(ptr) (((u8)((ptr)[0])))
+#define BSScriptReadPtr(ptr) ((u8*) BSScriptRead32(ptr))
+
+//read via add
+#define BS2ScriptRead32(ptr) ((ptr)[0] + ((ptr)[1] << 8) + ((ptr)[2] << 16) + ((ptr)[3] << 24))
+#define BS2ScriptRead16(ptr) ((ptr)[0] + ((ptr)[1] << 8))
+#define BS2ScriptReadPtr(ptr) ((u8*) BS2ScriptRead32(ptr))
+
+#define ATLEAST_ONE(value)(value != 0 ? value : 1)
+#define ATLEAST_ONE2(value)(value == 0 ? 1 : value)
+#define ATLEAST_ONE_NEGATIVE(value)(value != 0 ? value : -1)
+#define ATLEAST_ONE_PTR(value)(*value != 0 ? *value : (*value = 1))
+#define ATLEAST_ONE_PTR_NEGATIVE(value)(*value != 0 ? *value : (*value = -1))
```

Comment:
I don't really like these `ATLEAST_ONE_` macros, as they obfuscate the code. I'd inline them into the code unless you can find a name that makes it obvious what the macro does.